bindings/cpu_profiler.cc

```diff
@@ -58,59 +58,135 @@ v8::CpuProfilingLoggingMode GetLoggingMode() {
+  MeasurementsTicker(uv_loop_t* loop) :
     period_ms(100),
-    isolate(v8::Isolate::GetCurrent()),
-    heap_stats(v8::HeapStatistics())
-  {
+    isolate(v8::Isolate::GetCurrent())
+    {
     uv_timer_init(loop, &timer);
     timer.data = this;
-  };
+  }
+
+  static void ticker(uv_timer_t*);
+  // Memory listeners
+  void heap_callback();
+  void add_heap_listener(std::string& profile_id, std::function<void(uint64_t, v8::HeapStatistics&)> cb);
+  void remove_heap_listener(std::string& profile_id, std::function<void(uint64_t, v8::HeapStatistics&)>& cb);
 
-  static void callback(uv_timer_t* handle);
-  void add_listener(const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)> cb);
-  void remove_listener(const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)>& cb);
+  // CPU listeners
+  void cpu_callback();
+  void add_cpu_listener(std::string& profile_id, std::function<void(uint64_t, double)> cb);
+  void remove_cpu_listener(std::string& profile_id, std::function<void(uint64_t, double)>& cb);
+
+  size_t listener_count();
 };
 
-void HeapStatisticsTimer::callback(uv_timer_t* handle) {
-  HeapStatisticsTimer* self = static_cast<HeapStatisticsTimer*>(handle->data);
-  self->isolate->GetHeapStatistics(&self->heap_stats);
+size_t MeasurementsTicker::listener_count() {
+  return heap_listeners.size() + cpu_listeners.size();
+}
+
+// Heap tickers
+void MeasurementsTicker::heap_callback() {
+  isolate->GetHeapStatistics(&heap_stats);
   uint64_t ts = uv_hrtime();
 
-  for (auto& listener : self->listeners) {
-    listener.second(handle, ts, self->heap_stats);
+  for (auto& listener : heap_listeners) {
+    listener.second(ts, heap_stats);
+  }
+}
+
+void MeasurementsTicker::add_heap_listener(std::string& profile_id, std::function<void(uint64_t, v8::HeapStatistics&)> cb) {
+  heap_listeners.emplace(profile_id, cb);
+
+  if (listener_count() == 1) {
+    uv_timer_set_repeat(&timer, period_ms);
+    uv_timer_start(&timer, ticker, 0, period_ms);
+  }
+}
+
+void MeasurementsTicker::remove_heap_listener(std::string& profile_id, std::function<void(uint64_t, v8::HeapStatistics&)>& cb) {
+  heap_listeners.erase(profile_id);
+
+  if (listener_count() == 0) {
+    uv_timer_stop(&timer);
+  }
+};
+
+// CPU tickers
+void MeasurementsTicker::cpu_callback() {
+  uint64_t ts = uv_hrtime();
+
+  uv_cpu_info_t* cpu = &cpu_stats;
+  int count;
+  int err = uv_cpu_info(&cpu, &count);
+  if (err) {
+    return;
+  }
+
+  uint64_t total = 0;
+  uint64_t idle_total = 0;
+
+  for (int i = 0; i < count; i++) {
+    uv_cpu_info_t* core = cpu + i;
+
+    total += core->cpu_times.user;
+    total += core->cpu_times.nice;
+    total += core->cpu_times.sys;
+    total += core->cpu_times.idle;
+    total += core->cpu_times.irq;
+
+    idle_total += core->cpu_times.idle;
+  }
+
+  double total_avg = total / count;
+  double idle_avg = idle_total / count;
+  double rate = idle_avg / total_avg;
```

Comment:
I don't quite understand this - aren't we trying to measure the opposite (the non-idle time)?

Reply:
Yup, this should have been flipped. Updated to 1 - idle_avg/total_avg